ctapipe/calib/camera/flatfield.py

```diff
@@ -0,0 +1,247 @@
+            Set to None if no configuration to pass.
+        tool : ctapipe.core.Tool
+            Tool executable that is calling this component.
+            Passes the correct logger to the component.
+            Set to None if no Tool to pass.
+        kwargs
+
+        """
+        super().__init__(config=config, tool=tool, **kwargs)
+
+        self.log.info("Used events statistics : %d", self.max_events)
+        self.count = 0
+
+    def _extract_charge(self, event, tel_id):
+        """
+        Extract the charge and the time from a calibration event
+
+        Parameters
+        ----------
+        event : general event container
+
+        tel_id : telescope id
+        """
+
+        waveforms = event.r0.tel[tel_id].waveform
+
+        # Clean waveforms
+        if self.cleaner:
+            cleaned = self.cleaner.apply(waveforms)
+        # do nothing
+        else:
+            cleaned = waveforms
+
+        # Extract charge and time
+        if self.extractor:
+            if self.extractor.requires_neighbours():
+                g = event.inst.subarray.tel[tel_id].camera
+                self.extractor.neighbours = g.neighbor_matrix_where
+
+            charge, peak_pos, window = self.extractor.extract_charge(cleaned)
+
+        # sum all the samples
+        else:
+            charge = cleaned.sum(axis=2)
+            peak_pos = np.argmax(cleaned, axis=2)
+
+        return charge, peak_pos
+
+    def calculate_relative_gain(self, event, tel_id):
+        """
+        calculate the relative flat field coefficients
+
+        Parameters
+        ----------
+        event : general event container
+
+        tel_id : telescope id for which we calculate the gain
+        """
+
+        # initialize the np array at each cycle
+        waveform = event.r0.tel[tel_id].waveform
+        trigger_time = event.r0.tel[tel_id].trigger_time
+        pixel_status = event.r0.tel[tel_id].pixel_status
+
+        if self.count == 0:
+            self.time_start = trigger_time
+
+            if waveform.shape[0] < self.n_channels:
+                self.n_channels = waveform.shape[0]
+
+            self.event_median = np.zeros((self.max_events, self.n_channels))
+
+            n_pix = waveform.shape[1]
+            self.trace_integral = np.zeros((self.max_events, self.n_channels, n_pix))
+            self.trace_time = np.zeros((self.max_events, self.n_channels, n_pix))
+            self.trace_mask = np.zeros((self.max_events, self.n_channels, n_pix))
+
+        # extract the charge of the event and the peak position (assumed as time for the moment)
+        integral, peakpos = self._extract_charge(event, tel_id)
+
+        # remember the charge
+        self.trace_integral[self.count] = integral
+
+        # remember the time
+        self.trace_time[self.count] = peakpos
+
+        # keep the mask of not working pixels (to be improved)
+        self.trace_mask[self.count] = [pixel_status == 0, pixel_status == 0]
```

Comment:
why do we need a list of 2x `pixel_status == 0` again? I forgot ... why is this duplication needed?

Reply:
Ah, got it, not needed at all .. can be replaced by:

    self.trace_mask[self.count, :] = pixel_status == 0